lib/instrumentation/modules/@aws-sdk/client-s3.js

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright Elasticsearch B.V. and other contributors where applicable.
+ * Licensed under the BSD 2-Clause License; you may not use this file except in
+ * compliance with the BSD 2-Clause License.
+ */
+
+'use strict'
+
+const constants = require('../../../constants')
+
+const TYPE = 'storage'
+const SUBTYPE = 's3'
+const elasticAPMStash = Symbol('elasticAPMStash')
+const COMMAND_NAME_RE = /^(\w+)Command$/
+
+/**
+ * TODO: this method may be shared with other instrumentations
+ * For a HeadObject API call, `context.commandName === 'HeadObjectCommand'`.
+ *
+ * @param {String} commandName
+ * @returns {String}
+ */
+function opNameFromCommandName (commandName) {
+  const match = COMMAND_NAME_RE.exec(commandName)
+  if (match) {
+    return match[1]
+  } else {
+    return '<unknown command>'
+  }
+}
+
+/**
+ * Gets the region from the ARN
+ *
+ * @param {String} s3Arn
+ * @returns {String}
+ */
+function regionFromS3Arn (s3Arn) {
+  return s3Arn.split(':')[3]
+}
+
+/**
+ * Return an APM "resource" string for the bucket, Access Point ARN, or Outpost
+ * ARN. ARNs are normalized to a shorter resource name.
+ * Known ARN patterns:
+ * - arn:aws:s3:<region>:<account-id>:accesspoint/<accesspoint-name>
+ * - arn:aws:s3-outposts:<region>:<account>:outpost/<outpost-id>/bucket/<bucket-name>
+ * - arn:aws:s3-outposts:<region>:<account>:outpost/<outpost-id>/accesspoint/<accesspoint-name>
+ *
+ * In general that is:
+ *    arn:$partition:$service:$region:$accountId:$resource
+ *
+ * This parses using the same "split on colon" used by the JavaScript AWS SDK v3.
+ * https://github.com/aws/aws-sdk-js-v3/blob/v3.18.0/packages/util-arn-parser/src/index.ts#L14-L37
+ *
+ * @param {String} bucket The bucket string
+ * @returns {String}
+ */
+function resourceFromBucket (bucket) {
+  let resource = null
+  if (bucket) {
+    resource = bucket
+    if (resource.startsWith('arn:')) {
+      resource = bucket.split(':').slice(5).join(':')
+    }
+  }
+  return resource
+}
+
+/**
+ * We do alias them to a local type
+ * @typedef {import('@aws-sdk/types').InitializeMiddleware} InitalizeMiddleware
+ * @typedef {import('@aws-sdk/types').FinalizeRequestMiddleware } FinalizeRequestMiddleware
+ * @typedef {import('@aws-sdk/types').InitializeHandlerOptions} InitializeHandlerOptions
+ * @typedef {import('@aws-sdk/types').FinalizeRequestHandlerOptions } FinalizeRequestHandlerOptions
+ *
+ * Then create our types
+ * @typedef {InitalizeMiddleware | FinalizeRequestMiddleware} S3Middleware
+ * @typedef {InitializeHandlerOptions | FinalizeRequestHandlerOptions} S3MiddlewareOptions
+
+ * @typedef {object} S3MiddlewareEntry
+ * @property {S3Middleware} middleware
+ * @property {import('@aws-sdk/types').HandlerOptions} options
+ */
```

Comment:
This is a lot of content. ;)  Do you find it provides some useful completion in your editor?

Reply:
The intent was to have this conversation. From this small experience I can tell
- it seems you cannot import several types at once (but haven't check all docs yet)
- autocompletion is good and worked out of the box
- makes more sense in types that will be exported, this was used internally

In that specific scenario I think it makes more sense to:
- move types to smithy-client (since they are generic to all clients)
- import and use in client-s3 
- future AWS instrumentations like sns & dynamodb will profit from them

I'm fine dropping these lines but I think in object we have more intensive use (Agent, Span, Transaction, ...) we might notice the benefits

Reply:
I did move it to the smithy client so we can experiment on next clients (sqs, sns, dynamodb) and see if it' suseful